Documentation/General/emergentArchitectureDesign/emergentArchitectureDesign.tex

```diff
@@ -191,3 +191,47 @@ \section{Concurrency}
 These seperate threads will get a specific task assigned, which has his local data to compute on.
 Deadlocks will be prevented because the spawned threads will be joined before acquiring locks in synchronized statements.
 This will happen when we persist the algorithm output in the database.
+
+\section{Design Patterns}
+
+Due to the usage of various libraries and frameworks, our codebase does not contain a lot of helper mechanisms.
+Therefore the number of design patterns implemented is lower, because we use a lot of patterns defined in the frameworks.
+
+The design patterns we use, as they are defined in the frameworks, are:
+
+\begin{enumerate}
+\item Lazy initialization (Hibernate):\\
+OneToMany-relations are fetched from the database when they are needed.
+This makes retrieval of entities that use embeddables faster. For example we are most of the time not interested in the \class{VAVector} of a \class{Room}, only when we need to update such a \class{Room}, which happens rarely.
+
+\item Object pooling (RestEasy, Guice):\\
+The http-connection to SoundCloud and database-connection to our PostgreSQL database are pooled for re-use.
+This way we only have to set up the connection once using said frameworks and reduce object retrieval time.
+The \class{DBModule} uses a JpaPersistModule which is injected by Guice, whereas the \class{SoundCloudAPIConnector} uses a RestEasyClient.
+
+\item Memento (JPA, Guice):\\
+All database updates and retrievals are transactions, which impelement the Memento design pattern.
+This makes sure that all database changes are ACID.
+The methods are annotated by @Transactional, which Guice injects the transaction rollbacks.
+Then JPA can perform the fail-safes whenever necessary.
+All extensions of \class{AbstractDAO} have transactions.
+\end{enumerate}
+
+The design patterns we implemented ourselves are:
+
+\begin{enumerate}
+\item Singleton:\\
+In order to enforce a single controlling unit, all of our backends are @Singleton.
+For example the \class{ChatBackend} contains all roomInstances, which should be stateful in order to maintain cached collections of \class{ChatMessage}s.
```

Comment:
I think the fundamental requirement of REST is that it is stateless in the request point of view. Which means, that the previous request should not influence on *how* the next request is handled. Think of software setup gui's with next steps, that would **not** be stateless. The state should be stored on the client side (a session cookie), which theoretically should allow any of our servers to respond to the request. I mean, with messages comming in every millisecond, ofcourse we're switching some sort of state :wink: